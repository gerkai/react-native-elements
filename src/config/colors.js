import { StyleSheet } from 'react-native';

export default {
  primary: '#2089dc',
  secondary: '#8F0CE8',
  grey0: '#393e42',
  grey1: '#43484d',
  grey2: '#5e6977',
  grey3: '#86939e',
  grey4: '#bdc6cf',
  grey5: '#e1e8ee',
  greyOutline: '#bbb',
  searchBg: '#303337',
  error: '#ff190c',
  disabled: 'hsl(208, 8%, 90%)',
  // Darker color if hairlineWidth is not thin enough
  divider: StyleSheet.hairlineWidth < 1 ? '#bcbbc1' : 'rgba(0, 0, 0, 0.12)',
};
