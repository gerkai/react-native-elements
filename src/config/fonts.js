export default {
  ios: {
  },
  android: {
    regular: 'sans-serif',
    light: 'sans-serif-light',
    condensed: 'sans-serif-condensed',
    condensed_light: 'sans-serif-condensed-light',
    black: 'sans-serif-black',
    thin: 'sans-serif-thin',
    medium: 'sans-serif-medium',
    bold: 'sans-serif-bold',
  }
}
