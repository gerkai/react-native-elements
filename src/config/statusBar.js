import { Dimensions, Platform, StatusBar } from 'react-native';

// Borrowed from https://github.com/ptelad/react-native-iphone-x-helper

function isIphoneX() {
  const dimen = Dimensions.get('window');
  return (
    Platform.OS === 'ios' &&
    !Platform.isPad &&
    !Platform.isTVOS &&
    (dimen.height === 812 || dimen.width === 812)
  );
}

function ifIphoneX(iphoneXStyle, regularStyle) {
  if (isIphoneX()) {
    return iphoneXStyle;
  }
  return regularStyle;
}

export function getStatusBarHeight(skipAndroid = false) {
  if (Platform.OS === 'ios') {
    return ifIphoneX(44, 20);
  }

  if (skipAndroid) {
    return 0;
  }

  return StatusBar.currentHeight;
}
