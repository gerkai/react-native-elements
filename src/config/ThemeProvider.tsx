import React from 'react';
import deepmerge from 'deepmerge';
import colors from './colors';
import darkColors from './colorsDark';
import { FullTheme, Theme } from './theme';

type RecursivePartial<T> = { [P in keyof T]?: RecursivePartial<T[P]> };

export interface ThemeProps<T> {
  theme: Theme<T>;
  updateTheme: (updates: RecursivePartial<FullTheme>) => void;
  replaceTheme: (updates: RecursivePartial<FullTheme>) => void;
}

export const ThemeContext: React.Context<ThemeProps<{}>> = React.createContext({
  theme: {
    colors,
  },
} as ThemeProps<{}>);

export type ThemeProviderProps = {
  useDark?: boolean;
};

type ThemeProviderState = {
  theme: Theme;
  useDark: boolean;
};

export default class ThemeProvider extends React.Component<
  ThemeProviderProps,
  ThemeProviderState
> {
  static defaultProps = {
    theme: {},
    useDark: false,
  };
  defaultTheme: Partial<FullTheme>;

  constructor(props: { theme: Theme; useDark?: boolean }) {
    super(props);
    const defaultColors = props.useDark ? darkColors : colors;
    this.defaultTheme = deepmerge(
      {
        colors: defaultColors,
      },
      props.theme
    );
    this.state = {
      theme: this.defaultTheme,
      useDark: Boolean(props.useDark),
    };
  }

  static getDerivedStateFromProps(
    props: {
      theme: Theme;
      useDark?: boolean;
    },
    state: ThemeProviderState
  ) {
    const { useDark } = props;
    if (useDark !== state.useDark) {
      const defaultColors = useDark ? darkColors : colors;
      return {
        theme: deepmerge(
          state.theme,
          deepmerge(
            {
              colors: defaultColors,
            },
            props.theme
          )
        ),
        useDark,
      };
    }
    return null;
  }

  updateTheme = (updates: RecursivePartial<FullTheme>) => {
    this.setState(({ theme }) => ({
      theme: deepmerge(theme, updates),
    }));
  };

  replaceTheme = (theme: RecursivePartial<FullTheme>) => {
    this.setState(() => ({
      theme: deepmerge(this.defaultTheme, theme),
    }));
  };

  getTheme = () => this.state.theme;

  render() {
    return (
      <ThemeContext.Provider
        value={{
          theme: this.state.theme,
          updateTheme: this.updateTheme,
          replaceTheme: this.replaceTheme,
        }}
      >
        {this.props.children}
      </ThemeContext.Provider>
    );
  }
}

export const ThemeConsumer = ThemeContext.Consumer;
