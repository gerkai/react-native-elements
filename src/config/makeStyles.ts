import { useContext } from 'react';
import { StyleSheet } from 'react-native';
import { FullTheme } from './theme';
import { ThemeContext } from './ThemeProvider';

export const useTheme = () => {
  return useContext(ThemeContext);
};

export const makeStyles = <
  T extends StyleSheet.NamedStyles<T> | StyleSheet.NamedStyles<any>,
  V
>(
  styles: T | ((theme: Partial<FullTheme>, props: V) => T)
) => (props: V = {} as any): T => {
  const { theme } = useTheme();
  const css = typeof styles === 'function' ? styles(theme, props) : styles;

  return StyleSheet.create(css);
};
