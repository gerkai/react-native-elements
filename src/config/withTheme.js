import React from 'react';
import merge from 'lodash.merge';
import { ThemeConsumer } from './ThemeProvider';
import DefaultTheme from './theme';

const isClassComponent = Component =>
  Boolean(Component.prototype && Component.prototype.isReactComponent);

const withTheme = (WrappedComponent, themeKey) => {
  class ThemedComponent extends React.Component {
    render() {
      /* eslint-disable react/prop-types */
      const { forwardedRef, ...rest } = this.props;

      return (
        <ThemeConsumer>
          {context => {
            // If user isn't using ThemeProvider
            if (!context) {
              let props = { ...rest, theme: DefaultTheme };

              return isClassComponent(WrappedComponent) ? (
                <WrappedComponent ref={forwardedRef} {...props} />
              ) : (
                <WrappedComponent {...props} />
              );
            }

            const { theme, updateTheme } = context;
            const props = {
              theme,
              updateTheme,
              ...merge({}, themeKey && theme[themeKey], rest),
            };

            if (isClassComponent(WrappedComponent)) {
              return <WrappedComponent ref={forwardedRef} {...props} />;
            }
            return <WrappedComponent {...props} />;
          }}
        </ThemeConsumer>
      );
    }
  }

  const name = themeKey
    ? `Themed.${themeKey}`
    : `Themed.${WrappedComponent.displayName ||
        WrappedComponent.name ||
        'Component'}`;

  const forwardRef = (props, ref) => {
    return <ThemedComponent {...props} forwardedRef={ref} />;
  };

  ThemedComponent.displayName = name;
  forwardRef.displayName = name;

  // Forward refs from children
  if (isClassComponent(WrappedComponent)) {
    return React.forwardRef(forwardRef);
  }

  return ThemedComponent;
};

export default withTheme;
