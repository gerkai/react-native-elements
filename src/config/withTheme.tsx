import React from 'react';
import deepmerge from 'deepmerge';
import hoistNonReactStatics from 'hoist-non-react-statics';
import { ThemeConsumer, ThemeProps } from './ThemeProvider';
import DefaultTheme, { FullTheme } from './theme';

const isClassComponent = (Component: any) =>
  Boolean(Component.prototype && Component.prototype.isReactComponent);

export interface ThemedComponent {
  displayName: string;
}

const ThemedComponent = (
  WrappedComponent,
  themeKey?: string,
  displayName?: string
) => {
  return Object.assign(
    (props, forwardedRef) => {
      const { children, ...rest } = props;

      return (
        <ThemeConsumer>
          {(context) => {
            // If user isn't using ThemeProvider
            if (!context) {
              const newProps = { ...rest, theme: DefaultTheme, children };
              return isClassComponent(WrappedComponent) ? (
                <WrappedComponent ref={forwardedRef} {...newProps} />
              ) : (
                <WrappedComponent {...newProps} />
              );
            }
            const { theme, updateTheme, replaceTheme } = context;
            const newProps = {
              theme,
              updateTheme,
              replaceTheme,
              ...deepmerge<FullTheme>(
                (themeKey && theme[themeKey]) || {},
                rest,
                {
                  clone: false,
                }
              ),
              children,
            };
            if (isClassComponent(WrappedComponent)) {
              return <WrappedComponent ref={forwardedRef} {...newProps} />;
            }
            return <WrappedComponent {...newProps} />;
          }}
        </ThemeConsumer>
      );
    },
    { displayName: displayName }
  );
};

function withTheme<P = {}, T = {}>(
  WrappedComponent: React.ComponentType<P & ThemeProps<T>>,
  themeKey: string
):
  | React.FunctionComponent<Omit<P, keyof ThemeProps<T>>>
  | React.ForwardRefExoticComponent<P> {
  const name = themeKey
    ? `Themed.${themeKey}`
    : `Themed.${
        WrappedComponent.displayName || WrappedComponent.name || 'Component'
      }`;
  const Component = ThemedComponent(WrappedComponent, themeKey, name);

  if (isClassComponent(WrappedComponent)) {
    return hoistNonReactStatics(React.forwardRef(Component), WrappedComponent);
  }
  return Component;
}

export default withTheme;
