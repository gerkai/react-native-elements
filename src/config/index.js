import { Text } from 'react-native';
import BackgroundImage from './BackgroundImage';
import colors from './colors';
import { getStatusBarHeight } from './statusBar';
import ViewPropTypes from './ViewPropTypes';
import fonts from './fonts';
import ThemeProvider, { ThemeConsumer } from './ThemeProvider';
import merge from 'lodash.merge';
import withTheme from './withTheme';

const TextPropTypes = Text.propTypes;

export {
  BackgroundImage,
  colors,
  getStatusBarHeight,
  ViewPropTypes,
  TextPropTypes,
  fonts,
  ThemeProvider,
  ThemeConsumer,
  merge,
  withTheme,
};
