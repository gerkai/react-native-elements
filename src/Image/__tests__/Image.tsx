import React from 'react';
import { Image as RNImage } from 'react-native';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { create } from 'react-test-renderer';
import { ThemeProvider } from '../../config';
import ThemedImage, { Image } from '../index';

jest.useFakeTimers();

describe('Image Component', () => {
  beforeAll(() => {
    // useNativeDriver isn't available in jest, so just silencing the warning
    global.console.warn = () => null;
  });

  it('should render the appropriate testId when one is passed.', () => {
    const component = shallow(
      <Image
        testID="customTestId"
        source={{ uri: 'https://i.imgur.com/0y8Ftya.jpg' }}
      />
    );
    const image = component.find({ testID: 'customTestId' });
    expect(image.length).toBe(1);
  });

  it('should render the appropriate testId when one is not passed.', () => {
    const component = shallow(
      <Image source={{ uri: 'https://i.imgur.com/0y8Ftya.jpg' }} />
    );
    const image = component.find({ testID: 'RNE__Image' });
    expect(image.length).toBe(1);
  });

  it('should apply values from theme', () => {
    const theme = {
      Image: {
        placeholderStyle: {
          backgroundColor: 'red',
        },
      },
    };
    const component = create(
      <ThemeProvider theme={theme}>
        <ThemedImage source={{ uri: 'https://i.imgur.com/0y8Ftya.jpg' }} />
      </ThemeProvider>
    );
    expect(
      component.root.findByProps({ testID: 'RNE__Image__placeholder' }).props
        .style.backgroundColor
    ).toBe('red');
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should render without the transition', () => {
    const component = shallow(
      <Image
        source={{ uri: 'https://i.imgur.com/0y8Ftya.jpg' }}
        transition={false}
      />
    );
    component.find({ testID: 'RNE__Image' }).prop('onLoad')();
    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should apply value from style prop', () => {
    const component = shallow(
      <Image
        source={{ uri: 'https://i.imgur.com/0y8Ftya.jpg' }}
        style={{ tintColor: 'red' }}
      />
    );
    expect(component.find({ testID: 'RNE__Image' }).props().style).toEqual(
      expect.objectContaining({
        tintColor: 'red',
      })
    );
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should apply value from childrenContainerStyle prop', () => {
    const component = shallow(
      <Image
        source={{ uri: 'https://i.imgur.com/0y8Ftya.jpg' }}
        style={{ tintColor: 'red' }}
        childrenContainerStyle={{
          borderWidth: 1,
          borderColor: 'red',
        }}
      />
    );
    expect(component.find({ testID: 'RNE__Image' }).props().style).toEqual(
      expect.objectContaining({
        tintColor: 'red',
      })
    );

    expect(
      component.find({ testID: 'RNE__Image__children__container' }).props()
        .style
    ).toEqual(
      expect.objectContaining({
        borderWidth: 1,
        borderColor: 'red',
      })
    );

    expect(toJson(component)).toMatchSnapshot();
  });

  it('static methods should be present', () => {
    expect(Image.getSize).toBe(RNImage.getSize);
    expect(Image.getSizeWithHeaders).toBe(RNImage.getSizeWithHeaders);
    expect(Image.prefetch).toBe(RNImage.prefetch);
    expect(Image.abortPrefetch).toBe(RNImage.abortPrefetch);
    expect(Image.queryCache).toBe(RNImage.queryCache);
    expect(Image.resolveAssetSource).toBe(RNImage.resolveAssetSource);
  });
});
