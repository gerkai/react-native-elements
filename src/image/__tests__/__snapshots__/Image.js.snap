// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image Component should apply value from childrenContainerStyle prop 1`] = `
<View
  accessibilityIgnoresInvertColors={true}
  style={
    Object {
      "backgroundColor": "transparent",
      "overflow": "hidden",
      "position": "relative",
    }
  }
>
  <Image
    onLoad={[Function]}
    source={
      Object {
        "uri": "https://i.imgur.com/0y8Ftya.jpg",
      }
    }
    style={
      Object {
        "bottom": 0,
        "height": undefined,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "tintColor": "red",
        "top": 0,
        "width": undefined,
      }
    }
    testID="RNE__Image"
    transition={true}
    transitionDuration={360}
  />
  <ForwardRef(AnimatedComponentWrapper)
    accessibilityElementsHidden={true}
    importantForAccessibility="no-hide-descendants"
    pointerEvents="none"
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "opacity": 1,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#bdbdbd",
          "justifyContent": "center",
          "tintColor": "red",
        }
      }
      testID="RNE__Image__placeholder"
    />
  </ForwardRef(AnimatedComponentWrapper)>
  <View
    style={
      Object {
        "borderColor": "red",
        "borderWidth": 1,
      }
    }
    testID="RNE__Image__children__container"
  />
</View>
`;

exports[`Image Component should apply value from style prop 1`] = `
<View
  accessibilityIgnoresInvertColors={true}
  style={
    Object {
      "backgroundColor": "transparent",
      "overflow": "hidden",
      "position": "relative",
    }
  }
>
  <Image
    onLoad={[Function]}
    source={
      Object {
        "uri": "https://i.imgur.com/0y8Ftya.jpg",
      }
    }
    style={
      Object {
        "bottom": 0,
        "height": undefined,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "tintColor": "red",
        "top": 0,
        "width": undefined,
      }
    }
    testID="RNE__Image"
    transition={true}
    transitionDuration={360}
  />
  <ForwardRef(AnimatedComponentWrapper)
    accessibilityElementsHidden={true}
    importantForAccessibility="no-hide-descendants"
    pointerEvents="none"
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "opacity": 1,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#bdbdbd",
          "justifyContent": "center",
          "tintColor": "red",
        }
      }
      testID="RNE__Image__placeholder"
    />
  </ForwardRef(AnimatedComponentWrapper)>
  <View
    style={
      Object {
        "tintColor": "red",
      }
    }
    testID="RNE__Image__children__container"
  />
</View>
`;

exports[`Image Component should apply values from theme 1`] = `
<View
  accessibilityIgnoresInvertColors={true}
  style={
    Object {
      "backgroundColor": "transparent",
      "overflow": "hidden",
      "position": "relative",
    }
  }
>
  <Image
    onLoad={[Function]}
    replaceTheme={[Function]}
    source={
      Object {
        "uri": "https://i.imgur.com/0y8Ftya.jpg",
      }
    }
    style={
      Object {
        "bottom": 0,
        "height": undefined,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "width": undefined,
      }
    }
    testID="RNE__Image"
    theme={
      Object {
        "Image": Object {
          "placeholderStyle": Object {
            "backgroundColor": "red",
          },
        },
        "colors": Object {
          "black": "#242424",
          "disabled": "hsl(208, 8%, 90%)",
          "divider": "#bcbbc1",
          "error": "#ff190c",
          "grey0": "#393e42",
          "grey1": "#43484d",
          "grey2": "#5e6977",
          "grey3": "#86939e",
          "grey4": "#bdc6cf",
          "grey5": "#e1e8ee",
          "greyOutline": "#bbb",
          "platform": Object {
            "android": Object {
              "error": "#f44336",
              "grey": "rgba(0, 0, 0, 0.54)",
              "primary": "#2196f3",
              "searchBg": "#dcdce1",
              "secondary": "#9C27B0",
              "success": "#4caf50",
              "warning": "#ffeb3b",
            },
            "default": Object {
              "error": "#ff3b30",
              "grey": "#7d7d7d",
              "primary": "#007aff",
              "searchBg": "#dcdce1",
              "secondary": "#5856d6",
              "success": "#4cd964",
              "warning": "#ffcc00",
            },
            "ios": Object {
              "error": "#ff3b30",
              "grey": "#7d7d7d",
              "primary": "#007aff",
              "searchBg": "#dcdce1",
              "secondary": "#5856d6",
              "success": "#4cd964",
              "warning": "#ffcc00",
            },
            "web": Object {
              "error": "#ff190c",
              "grey": "#393e42",
              "primary": "#2089dc",
              "searchBg": "#303337",
              "secondary": "#ca71eb",
              "success": "#52c41a",
              "warning": "#faad14",
            },
          },
          "primary": "#2089dc",
          "searchBg": "#303337",
          "secondary": "#ca71eb",
          "success": "#52c41a",
          "warning": "#faad14",
          "white": "#ffffff",
        },
      }
    }
    transition={true}
    transitionDuration={360}
    updateTheme={[Function]}
  />
  <View
    accessibilityElementsHidden={true}
    importantForAccessibility="no-hide-descendants"
    pointerEvents="none"
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "opacity": 1,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "red",
          "justifyContent": "center",
        }
      }
      testID="RNE__Image__placeholder"
    />
  </View>
  <View
    style={Object {}}
    testID="RNE__Image__children__container"
  />
</View>
`;

exports[`Image Component should render without the transition 1`] = `
<View
  accessibilityIgnoresInvertColors={true}
  style={
    Object {
      "backgroundColor": "transparent",
      "overflow": "hidden",
      "position": "relative",
    }
  }
>
  <Image
    onLoad={[Function]}
    source={
      Object {
        "uri": "https://i.imgur.com/0y8Ftya.jpg",
      }
    }
    style={
      Object {
        "bottom": 0,
        "height": undefined,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "width": undefined,
      }
    }
    testID="RNE__Image"
    transition={false}
    transitionDuration={360}
  />
  <ForwardRef(AnimatedComponentWrapper)
    accessibilityElementsHidden={true}
    importantForAccessibility="no-hide-descendants"
    pointerEvents="none"
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "opacity": 0,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#bdbdbd",
          "justifyContent": "center",
        }
      }
      testID="RNE__Image__placeholder"
    />
  </ForwardRef(AnimatedComponentWrapper)>
  <View
    style={Object {}}
    testID="RNE__Image__children__container"
  />
</View>
`;
